wire-ios-sync-engine/Source/UserSession/URLActionProcessors/DeepLinkURLActionProcessor.swift

```diff
@@ -36,96 +36,129 @@ class DeepLinkURLActionProcessor: URLActionProcessor {
             }
+        }
+    }
 
-        case .openConversation(let id):
-            let viewContext = contextProvider.viewContext
-            guard let conversation = ZMConversation.fetch(with: id, domain: nil, in: viewContext) else {
-                delegate?.failedToPerformAction(urlAction, error: DeepLinkRequestError.invalidConversationLink)
+    private func handlePasswordPrompt(for conversationName: String, key: String, code: String, delegate: PresentationDelegate) {
+        delegate.showPasswordPrompt(for: conversationName) { [weak self] password in
+            guard let self, let password = password, !password.isEmpty else {
                 return
             }
 
-            delegate?.showConversation(conversation, at: nil)
-            delegate?.completedURLAction(urlAction)
+            self.joinConversation(key: key, code: code, password: password, delegate: delegate)
+        }
+    }
 
-        case .openUserProfile(let id):
-            let viewContext = contextProvider.viewContext
-            if let user = ZMUser.fetch(with: id, domain: nil, in: viewContext) {
-                delegate?.showUserProfile(user: user)
-            } else {
-                delegate?.showConnectionRequest(userId: id)
+    private func handleJoinWithoutPassword(for conversationName: String, key: String, code: String, urlAction: URLAction, delegate: PresentationDelegate) {
+        delegate.shouldPerformActionWithMessage(conversationName, action: .joinConversation(key: key, code: code)) { [weak self] shouldJoin in
+            guard let self, shouldJoin else {
+                delegate.completedURLAction(urlAction)
+                return
             }
 
-            delegate?.completedURLAction(urlAction)
+            self.joinConversation(key: key, code: code, password: nil, delegate: delegate)
+        }
+    }
 
-        default:
-            delegate?.completedURLAction(urlAction)
+    private func joinConversation(key: String, code: String, password: String?, delegate: PresentationDelegate) {
+        ZMConversation.join(
+            key: key,
+            code: code,
+            password: password,
+            transportSession: transportSession,
+            eventProcessor: eventProcessor,
+            contextProvider: contextProvider
+        ) { [weak self] response in
+            guard let self else { return }
+
+            switch response {
+            case .success(let conversation):
+                self.synchronise(conversation) { result in
+                    DispatchQueue.main.async {
+                        switch result {
+                        case .success(let syncConversation):
+                            delegate.showConversation(syncConversation, at: nil)
+                        case .failure(let error):
+                            delegate.failedToPerformAction(.joinConversation(key: key, code: code), error: error)
+                        }
+                    }
+                }
```

Comment:
Could this be extracted to new func?

Reply:
Done: https://github.com/wireapp/wire-ios/commit/fe73ec21eb1b0edff09974ac2101279f381944d1